### DISCUSSION TITLE
Global variant options and variant generator

### DISCUSSION THREAD

**Discussion by genox** [medusajs/medusa#5119] [Feature Requests]
I happen to work on a store for clothing. Clothing with lots of sizes and colours.

While evaluating other stores including medusa, I saw that medusa neither supports global attributes used for variant creation nor does it offer a variant generator per se. This would be enormously practical to enter data in an efficient manner.

In a nutshell, most solutions I checked out handle it like this:

1) a global attribute list, e.g. to manage size and color names in one place
2) during variant creation, one selects the attributes and their values that the product is available in (red, black; xl, l, s, m)
3) the generator generates a list of all possible combinations with SKU fields that contain some unique ID that can be replaced with a SKU

**Comment by pevey** [comment]
Magento handles variants similarly.  You could set up different attributes and then also set up an attribute set of one or more attributes, and then assign an attribute set to a product.  This was extremely handy not only in saving time, but also improving consistency/reducing errors.  It can also help with reporting on sales by attribute.

It is probably the one and only thing I actually miss from Magento.

**Reply by voxsar** [reply]
For some reason the v1 did not have the ability to set metadata for product options and their values, I have created a plugin that allows to add such metadata

Use Cases, lets say you want to store the option's image or the option's values image like

E.g. A shoe product with color option, each option value can have meta data for a variant image (red, blue etc), (which can also be done using variant images, but it's just an example)

Another example is using the meta data to inform how the react front end should render the options, a select box, radio buttons

/admin/products/{id}/options/{id}/option

{
metadata: {
   inputtype: 'select'
}

I made this plugin that extends the functionality of the medusa 1.x to allow adding metadata to options and values

**Comment by tsvetann** [comment]
I like this feature

**Comment by genox** [comment]
The headless commerce solutions I checked out also for some reason have trouble providing color swatch attribute lists, which is one of the most basic requirements for many stores that offer configurable products.

You might want to check out the way swell.is is handling this, it's the perfect template. Saleor offers the same and has a more advanced variant editor (essentially editing a spreadsheet in the browser with copy/paste of cells, very convenient for 10+ variants)

Attribute lists are just nice. In frontends, we always have to map and hardcode user entered data to UI select options. A typo entered in the product admin UI can literally brick a frontend if not handled correctly (lets just assume the "real world" scenario). Offering predefined, controlled attribute lists (key/value pairs) essentially completely solves that issue.

As soon as the use case grows into a little bit more complex needs, you will face issues like this and they are - in my naive world - easily solvable. :)

Since you will be going down the route of generating OpenAPI clients at some point, those same attribute lists can become an option to type properties on products for even more easier development support. Something you will never have on a field typed "string".

**Comment by Frioo** [comment]
+1 for sure! If I remember correctly, even GoDaddy WYSWIG e-commerce creator allows specifying a global list of options and variant generation. Would be great to see this in medusa. Also optional thumbnails/swatches for option values. Currently I'd have to resort to a separate CMS to store these additional presentation fields.

**Comment by HarleySalas** [comment]
Has anyone come up with a solution or plugin for this? It's something I need in every single project. Really surprised this hasn't gotten more demand/attention.

**Comment by chiuwah** [comment]
@genox What solution did you find for your fashion store? Were you able to customize Medusa? I am looking for the same feature for my project. I stumbled on this plugin: https://github.com/vholik/medusa-custom-attributes However, it is only on the product level, and not on the variant level, when creating a product. 

I created a test account on swell.is, but could not find the mentioned template function. Could you maybe share some more information how swell.is implented this function?

**Reply by genox** [reply]
Since my project depends on an import workflow for products, I save all variant information into metadata. It is not exactly convenient to administer it in the admin UI for obvious reasons like this. But it reduces code complexity by a lot.

I hope 2.0 adds this out of the box or makes it easier.

**Reply by chiuwah** [reply]
Yes, I guess the metadata is the best solution for now, without having to extend the model and services. One thing that comes to my mind is how performance with search and filtering on metadata is?

**Reply by genox** [reply]
To reduce API requests and circumvent limitations on filtering/search, feature wise and speed, I customised the meilisearch index based on the available plugi. data is only directly grabbed from the API to get specific products and their variations, e.g. the detail views.

**Comment by Pigamo** [comment]
I agree with this post. I was going to write something similar. Options should build up to a variant. It's extremely inconvenient to have to retype the variant option for each one. 

This also makes filtering more challenging as now I have to match titles and build it up from there.

Options should be global and variants generated!

**Comment by NicolasGorga** [comment]
Hello, has anyone found a viable alternative for this in v1?

**Comment by qFlensT** [comment]
+1 in 2025

**Comment by totoLab** [comment]
Currently considering this project as a Magento2 replacement, this is the only missing feature I would need. I invite you to consider the case when a product must be compatible with something, let's say in a phone cases business: you have multiple cases that differ in design or some other feature and the user needs to select which iPhone they have from a list. There would not be so many items in the dropdown but you still need to update the list and every new product should have the updated list associated. I think it's a vital feature. 
In Magento2 I'm currently using Mageworx Option Templates extension to achieve this.

**Comment by imash96** [comment]
#13817 wip

**Comment by MicahBCode** [comment]
Hello guys,
As I‘ve recently checked out the current progress on this feature in the current PRs I would like to share my thoughts. The PRs show a differentation between global and exclusive product options on a product options level.

> This means that those product options will not be treatede as Global, which are the ones a user could filter by in a storefront or can be linked to other products.

This description leaves me wondering why this separation between "Global" and "Exclusive" product options is needed in the first place. From my perspective, removing this distinction whould actually increase flexibility.

Many ecommerce setups define "product option requirements" at the category level - meaning certain categories require or optionally support specific product options. In these cases, the storefront filters should always reflect only the product options that appear within the products of the currently viewed category or page. They shouldn't be globally static across the entire storefront.

That's why I'm unsure about the value of distinguishing between "Global" and "Exclusive" options.
If the idea is that

- Global options = options intended for storefront filtering
- Exclusive options = specifications that shouldn't appear as filters

... then that distinction would arguably make more sense if it were configurable per category, rather than on a global level as in some categories, I might want product option X to be intended for filtering in the storefront but not in other categories.

Could you clarify the intention behind introducing this separation? And wouldn't it make sense to make it configurable on a category level instead?

@willbouch @olivermrbl Could you share some insights about this decision and your thoughts about my suggested changes?

**Comment by danielgran** [comment]
That would be super cool.
At that point I'm missing a product pricing tool where we can (not via price lists) configure the prices for all the product variants haha. The grid even would support that
